File: poc-wasm/hosts/node/benchmark.mjs
import { readFile } from 'fs/promises';
import { WASI } from 'wasi';
import { argv, env } from 'process';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { performance } from 'perf_hooks';

const __dirname = dirname(fileURLToPath(import.meta.url));

const WARMUP_RUNS = 3;
const BENCHMARK_RUNS = 10;

async function loadModule(name) {
  const wasmPath = join(__dirname, `../../build/lumis-lang-${name}.wasm`);
  const wasmBuffer = await readFile(wasmPath);

  const wasi = new WASI({
    version: 'preview1',
    args: argv,
    env,
  });

  const module = await WebAssembly.compile(wasmBuffer);
  const instance = await WebAssembly.instantiate(module, {
    wasi_snapshot_preview1: wasi.wasiImport,
  });

  wasi.initialize(instance);

  return instance.exports;
}

async function benchmarkLoad(name) {
  // Warmup
  for (let i = 0; i < WARMUP_RUNS; i++) {
    await loadModule(name);
  }

  // Measured runs
  const times = [];
  for (let i = 0; i < BENCHMARK_RUNS; i++) {
    const start = performance.now();
    await loadModule(name);
    const elapsed = performance.now() - start;
    times.push(elapsed);
  }

  const mean = times.reduce((a, b) => a + b, 0) / times.length;
  return mean;
}

async function main() {
  console.log('=== Lumis WASM Benchmark - Node.js ===');
  console.log();
  console.log(`Warmup: ${WARMUP_RUNS}, Runs: ${BENCHMARK_RUNS}`);
  console.log();

  const languages = [
    'html', 'css', 'javascript', 'typescript', 'json',
    'rust', 'go', 'c',
    'python', 'ruby', 'bash', 'lua'
  ];

  const bundles = [
    'bundle-web',
    'bundle-system',
    'bundle-scripting'
  ];

  console.log('--- Individual Languages ---');
  console.log('| Language   | Mean (ms) |');
  console.log('|------------|-----------|');

  for (const lang of languages) {
    const mean = await benchmarkLoad(lang);
    console.log(`| ${lang.padEnd(10)} | ${mean.toFixed(2).padStart(9)} |`);
  }

  console.log();
  console.log('--- Bundles ---');
  console.log('| Bundle           | Mean (ms) |');
  console.log('|------------------|-----------|');

  for (const bundle of bundles) {
    const mean = await benchmarkLoad(bundle);
    console.log(`| ${bundle.padEnd(16)} | ${mean.toFixed(2).padStart(9)} |`);
  }
}

main().catch(console.error);

File: poc-wasm/hosts/java/src/main/java/dev/lumis/Benchmark.java
package dev.lumis;

import com.dylibso.chicory.runtime.Store;
import com.dylibso.chicory.wasi.WasiOptions;
import com.dylibso.chicory.wasi.WasiPreview1;
import com.dylibso.chicory.wasm.Parser;

import java.nio.file.Path;

public class Benchmark {
    private static final int WARMUP_RUNS = 3;
    private static final int BENCHMARK_RUNS = 10;

    private final String baseDir;

    public Benchmark() {
        this.baseDir = System.getProperty("user.dir");
    }

    private void loadModule(String name) throws Exception {
        Path wasmPath = Path.of(baseDir, "../../build/lumis-lang-" + name + ".wasm");

        var wasiOptions = WasiOptions.builder().build();
        var wasi = WasiPreview1.builder().withOptions(wasiOptions).build();

        var module = Parser.parse(wasmPath.toFile());

        var store = new Store().addFunction(wasi.toHostFunctions());
        store.instantiate(name, module);
    }

    private double benchmarkLoad(String name) throws Exception {
        // Warmup
        for (int i = 0; i < WARMUP_RUNS; i++) {
            loadModule(name);
        }

        // Measured runs
        long total = 0;
        for (int i = 0; i < BENCHMARK_RUNS; i++) {
            long start = System.nanoTime();
            loadModule(name);
            total += System.nanoTime() - start;
        }

        return (double) total / BENCHMARK_RUNS / 1_000_000.0;
    }

    public static void main(String[] args) throws Exception {
        System.out.println("=== Lumis WASM Benchmark - Java (Chicory) ===");
        System.out.println();
        System.out.printf("Warmup: %d, Runs: %d%n%n", WARMUP_RUNS, BENCHMARK_RUNS);

        Benchmark bench = new Benchmark();

        String[] languages = {
            "html", "css", "javascript", "typescript", "json",
            "rust", "go", "c",
            "python", "ruby", "bash", "lua"
        };

        String[] bundles = {
            "bundle-web",
            "bundle-system",
            "bundle-scripting"
        };

        System.out.println("--- Individual Languages ---");
        System.out.println("| Language   | Mean (ms) |");
        System.out.println("|------------|-----------|");

        for (String lang : languages) {
            double mean = bench.benchmarkLoad(lang);
            System.out.printf("| %-10s | %9.2f |%n", lang, mean);
        }

        System.out.println();
        System.out.println("--- Bundles ---");
        System.out.println("| Bundle           | Mean (ms) |");
        System.out.println("|------------------|-----------|");

        for (String bundle : bundles) {
            double mean = bench.benchmarkLoad(bundle);
            System.out.printf("| %-16s | %9.2f |%n", bundle, mean);
        }
    }
}
